[MRG+1] Rename CV params n_{folds,iter} to n_splits

yenchenlin opened the pull request:
Since now all CV splitters have `get_n_splits()`, we can rename n_{folds,iter} to n_splits in order to avoid confusion.

More detailed discussion can be found in #7169 .
- [x] Rename `n_iter` to `n_splits`
- [x] Rename `n_ folds` to `n_splits`

yenchenlin committed:
Rename n_iter to n_splits

yenchenlin committed:
Fix bug

jnothman commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -134,7 +134,7 @@ def test_cross_validator_with_default_params():
     n_folds = 2
     p = 2
-    n_iter = 10  # (the default value)
+    n_splits= 10  # (the default value)
```
space

jnothman commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -653,15 +653,15 @@ def test_label_shuffle_split():
     for l in labels:
         X = y = np.ones(len(l))
-        n_iter = 6
+        n_splits= 6
```
space

jnothman commented:
One question is whether we should be doing any deprecation... I think not, in this case.

yenchenlin committed:
Fix examples

yenchenlin committed:
Add spaces

yenchenlin committed:
Rename n_folds to n_splits

jnothman commented:
Test failures.

yenchenlin commented:
Oh yeah, I submit this PR to ask question:

Should I modify old splitter in `sklearn/cross_validation.py`?
[This line](https://github.com/scikit-learn/scikit-learn/pull/7187/files#diff-3f55cc702cd72cfa9923bf58091924f3R878) seems to use old splitter, and therefore results in error.

yenchenlin committed:
Fix error

jnothman commented:
More test failures, though!

yenchenlin committed:
Fix doc

jnothman commented:
Remaining failures are uses of `n_folds` in doc/tutorial/statistical_inference/model_selection.rst

yenchenlin committed:
Fix doc

yenchenlin committed:
Fix example

yenchenlin commented:
@jnothman @MechCoder CI is smiling 😄 , I think it's ready to be reviewed!

jnothman commented on examples/svm/plot_svm_scale_c.py:
```diff
@@ -128,7 +128,7 @@
         # reduce the variance
         grid = GridSearchCV(clf, refit=False, param_grid=param_grid,
-                            cv=ShuffleSplit(train_size=train_size, n_iter=250,
+                            cv=ShuffleSplit(train_size=train_size, n_splits=250,
```
PEP8

raghavrv commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -577,19 +578,19 @@ def assert_counts_are_ok(idx_counts, p):
         train_counts = [0] * n_samples
         test_counts = [0] * n_samples
-        n_splits = 0
+        splits_cnt = 0
```
`n_splits_test` maybe?

jnothman replied:
Or `n_splits_actual`

jnothman commented:
Multiple PEP8 violations...

jnothman commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -132,42 +132,43 @@ def get_params(self, deep=False):
 
-    n_splits = [n_samples, comb(n_samples, p), n_folds, n_folds,
-                n_unique_labels, comb(n_unique_labels, p), n_iter, 2]
+    splits_cnts = [n_samples, comb(n_samples, p), n_splits, n_splits,
```
`expected_n_splits` or `n_splits_expected` might be better.

jnothman replied:
The issue is that `cnt`, apart from being an abbreviation we should try to avoid for clarity, means the same as `n`. So `splits_cnt` means the same as `n_splits` and so the naming tells us little of what it contains.

Another option is renaming the new `n_splits` to `n_splits_param`

raghavrv replied:
I agree!

yenchenlin replied:
Thanks for the elaboration, it's useful for me in the future 😄

jnothman commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -562,7 +563,7 @@ def test_stratified_shuffle_split_even():
     # equal chance
     n_folds = 5
-    n_iter = 1000
+    n_splits = 1000
```
again, this could be `n_splits_param`

yenchenlin replied:
After modified [this line](https://github.com/scikit-learn/scikit-learn/pull/7187/files#diff-3f55cc702cd72cfa9923bf58091924f3R587), I guess `n_splits` can remain the same name?

raghavrv replied:
Yes. Leave it as such...

jnothman commented on sklearn/model_selection/_split.py:
```diff
@@ -566,34 +566,34 @@ def _make_test_folds(self, X, y=None, labels=None):
-        if np.all(self.n_folds > y_counts):
+        if np.all(self.n_splits > y_counts):
             raise ValueError("All the n_labels for individual classes"
                              " are less than %d folds."
```
maybe should be `less than n_splits=%d`, although this whole error message could do with better wording.

yenchenlin replied:
Thanks! I also fixed some other error messages.

jnothman commented:
Otherwise LGTM

jnothman commented:
Please draft a change to what's new, both in the `Model Selection Enhancements` section and in, perhaps, API Changes.

raghavrv commented:
Yes this should go into model selection enhancements section. Otherwise with a whatsnew and reviews addressed this LGTM too.... Let's merge this! Thanks for the PR @yenchenlin

yenchenlin committed:
PEP8

yenchenlin committed:
Rename variables name

yenchenlin committed:
Fix error message

yenchenlin committed:
Add whats_new

yenchenlin commented:
I've addressed the comments and also

1. check PEP8 length of each changed files
2. modify some error messages in `sklearn/model_selection/_split.py`

Thanks for the inputs!

yenchenlin committed:
Fix test

jnothman commented on doc/whats_new.rst:
```diff
@@ -62,6 +62,16 @@ Model Selection Enhancements and API Changes
     the corresponding parameter is not applicable. Additionally a list of all
     the parameter dicts are stored at ``results_['params']``.
 
+  - **Renaming of `n_folds` and `n_iter` to `n_splits`**
```
needs double backticks for code formatting, here and below.

jnothman commented on doc/whats_new.rst:
```diff
@@ -62,6 +62,16 @@ Model Selection Enhancements and API Changes
 
+  - **Renaming of `n_folds` and `n_iter` to `n_splits`**
+
+    The number of folds (i.e. the number of partitions of a dataset) is not 
```
I think this doesn't need so much motivation here, you just need to say something like "In the new `model_selection` module, some parameter names have changed: the `n_folds` parameter in x, y, z is now `n_splits`; the `n_iter` parameter in a, b, c is now `n_splits`". Please use `:class:` as appropriate.

jnothman commented on doc/whats_new.rst:
```diff
@@ -353,6 +363,11 @@ API changes summary
      (`#6697 <https://github.com/scikit-learn/scikit-learn/pull/6697>`_) by
      `Raghav R V`_.
 
+   - The parameters `n_iter` or `n_folds` in old CV splittersare are deprecated 
```
double backticks

jnothman replied:
*"splitters are"

jnothman replied:
Not "deprecated" as this usually refers to something that we still support.

yenchenlin committed:
Fix doc

yenchenlin commented:
Comments addressed, please have a look!

jnothman commented:
LGTM

raghavrv commented on doc/whats_new.rst:
```diff
@@ -62,6 +62,16 @@ Model Selection Enhancements and API Changes
 
+  - **Renaming of ``n_folds`` and ``n_iter`` to ``n_splits``**
+
+    Some parameter names have changed: 
```
You could remove this line I think. You already have a heading...

jnothman replied:
I think this line is good.

raghavrv replied:
I thought after [this](https://github.com/scikit-learn/scikit-learn/pull/7187#discussion_r74945171) change, this is not needed...

jnothman replied:
Perhaps...? I'm unconvinced that this series of comments markedly contributes to the quality of the text in terms of convention, clarity or information structure.

raghavrv commented on doc/whats_new.rst:
```diff
@@ -62,6 +62,16 @@ Model Selection Enhancements and API Changes
+    Some parameter names have changed: 
+    the ``n_folds`` parameter in :class:`model_selection.KFold`, 
+    :class:`model_selection.LabelKFold`, and 
+    :class:`model_selection.StratifiedKFold` is now ``n_splits``; 
```
is now renamed to

raghavrv replied:
Full stop and not a semi colon

raghavrv commented on doc/whats_new.rst:
```diff
@@ -62,6 +62,16 @@ Model Selection Enhancements and API Changes
+    the ``n_folds`` parameter in :class:`model_selection.KFold`, 
+    :class:`model_selection.LabelKFold`, and 
+    :class:`model_selection.StratifiedKFold` is now ``n_splits``; 
+    the ``n_iter`` in :class:`model_selection.ShuffleSplit`, 
```
The

raghavrv replied:
``n_iter`` parameter

raghavrv commented on doc/whats_new.rst:
```diff
@@ -62,6 +62,16 @@ Model Selection Enhancements and API Changes
     the corresponding parameter is not applicable. Additionally a list of all
     the parameter dicts are stored at ``results_['params']``.
 
+  - **Renaming of ``n_folds`` and ``n_iter`` to ``n_splits``**
```
**Parameters ``n_folds`` and ``n_iter`` renamed to ``n_splits``**?

raghavrv commented on doc/whats_new.rst:
```diff
@@ -62,6 +62,16 @@ Model Selection Enhancements and API Changes
+  - **Renaming of ``n_folds`` and ``n_iter`` to ``n_splits``**
+
+    Some parameter names have changed: 
+    the ``n_folds`` parameter in :class:`model_selection.KFold`, 
```
The

raghavrv commented on doc/whats_new.rst:
```diff
@@ -62,6 +62,16 @@ Model Selection Enhancements and API Changes
+    :class:`model_selection.StratifiedKFold` is now ``n_splits``; 
+    the ``n_iter`` in :class:`model_selection.ShuffleSplit`, 
+    :class:`model_selection.LabelShuffleSplit`, 
+    and :class:`model_selection.StratifiedShuffleSplit` is now ``n_splits``, too.
```
renamed to

raghavrv commented on doc/whats_new.rst:
```diff
@@ -353,6 +363,11 @@ API changes summary
+   - The parameters ``n_iter`` or ``n_folds`` in old CV splitters are replaced
+     by the new parameter ``n_splits`` since it can provide a consistent 
+     and unambiguous interface to represent the number of train-test splits.
+     By `YenChen Lin`_.
```
Add a link to the PR. It kind of gives a back reference from whatsnew to the PR / Issue...

raghavrv commented on sklearn/model_selection/tests/test_split.py:
```diff
@@ -616,7 +617,7 @@ def test_stratified_shuffle_split_overlap_train_test_bug():
     y = [0, 1, 2, 3] * 3 + [4, 5] * 5
     X = np.ones_like(y)
 
-    splits = StratifiedShuffleSplit(n_iter=1,
```
`splits` --> `sss`

(This is a bit unrelated to the PR... But since you are touching the code, you might as well change it...)

yenchenlin replied:
Thanks for pointing out!

raghavrv commented on sklearn/model_selection/tests/test_validation.py:
```diff
@@ -136,7 +136,8 @@ def _is_training_data(self, X):
 y = np.array([0, 0, 1, 1, 2, 2, 3, 3, 4, 4])
-# The number of samples per class needs to be > n_folds, for StratifiedKFold(3)
+# The number of samples per class needs to be > n_splits,
+# for StratifiedKFold(3)
```
pedantic nitpick: for `StratifiedKFold(n_splits=3)` to be explicit...

raghavrv commented:
Feel free to add a +1 once the comments are addressed. Ping @jnothman for merge once done...

raghavrv commented:
@jnothman Would you like to merge it as such? LGTM to me...

yenchenlin committed:
Fix doc

yenchenlin committed:
Make test clear

yenchenlin commented:
@raghavrv great thanks a lot for your review!
@jnothman comments addressed!

agramfort commented:
Lgtm

@jnothman merge if you are happy

MechCoder commented:
(Y)

amueller commented:
aaaand my book broke ^^ glad it's not in print yet lol. I should do that release some time soom.

jnothman commented:
hahahahaha. Glad you wrote it for the version where everything changes.

amueller commented:
Well that's why I knew I couldn't do it for 0.17. But It would be great to release 0.18 within the next month (which I think is doable).

amueller commented:
Shouldn't the default implementation of ``get_n_splits`` be ``return n_splits``?

jnothman commented:
I think it is better explicitly implemented...?
